superset-frontend/src/components/Datasource/DatasourceModal.tsx

```diff
@@ -202,11 +207,16 @@ const DatasourceModal: FunctionComponent<DatasourceModalProps> = ({
       })
       .catch(response => {
         setIsSaving(false);
-        getClientErrorObject(response).then(({ error }) => {
+        response.json().then((errorJson: ErrorResponse) => {
           modal.error({
-            title: t('Error'),
-            content: error || t('An error has occurred'),
+            title: t('Error saving dataset'),
             okButtonProps: { danger: true, className: 'btn-danger' },
+            content: (
+              <ErrorMessageWithStackTrace
+                error={errorJson.errors[0]}
```

Comment:
How confident are we that the api response will have this format?

Reply:
Because I removed the `@safe` decorator from the API endpoint, 100%. But someone could add the decorator back, so I'll add some safeguards here.